op-challenger/challenger/log_store_test.go

```diff
@@ -0,0 +1,225 @@
+	err := logStore.Subscribe()
+	require.NoError(t, err)
+
+	// Quit the subscription
+	logStore.Quit()
+	require.Nil(t, logStore.subscription)
+}
+
+// TestLogStore_Resubsribe tests the [Resubscribe] method on a [logStore].
+func TestLogStore_Resubscribe(t *testing.T) {
+	query := ethereum.FilterQuery{}
+	client := newMockLogStoreClient()
+	log := testlog.Logger(t, log.LvlError)
+	logStore := NewLogStore(query, client, log)
+
+	// Subscribe to the logStore.
+	err := logStore.Subscribe()
+	require.NoError(t, err)
+
+	// Resubscribe to the logStore.
+	err = logStore.resubscribe()
+	require.NoError(t, err)
+}
+
+// TestLogStore_Logs tests log methods on a [logStore].
+func TestLogStore_Logs(t *testing.T) {
+	query := ethereum.FilterQuery{}
+	client := newMockLogStoreClient()
+	log := testlog.Logger(t, log.LvlError)
+	logStore := NewLogStore(query, client, log)
+
+	require.Equal(t, []types.Log{}, logStore.GetLogs())
+	require.Equal(t, []types.Log(nil), logStore.GetLogByBlockHash(common.HexToHash("0x1")))
+
+	// Insert logs.
+	logStore.insertLog(types.Log{
+		BlockHash: common.HexToHash("0x1"),
+	})
+	logStore.insertLog(types.Log{
+		BlockHash: common.HexToHash("0x1"),
+	})
+
+	// Validate log insertion.
+	require.Equal(t, 2, len(logStore.GetLogs()))
+	require.Equal(t, 2, len(logStore.GetLogByBlockHash(common.HexToHash("0x1"))))
+}
+
+// TestLogStore_DispatchLogs tests the [DispatchLogs] method on the [logStore].
+func TestLogStore_DispatchLogs(t *testing.T) {
+	query := ethereum.FilterQuery{}
+	client := newMockLogStoreClient()
+	log := testlog.Logger(t, log.LvlError)
+	logStore := NewLogStore(query, client, log)
+
+	// Subscribe to the logStore.
+	err := logStore.Subscribe()
+	require.NoError(t, err)
+
+	// Dispatch logs on the logStore.
+	go logStore.dispatchLogs()
+
+	// Send logs through the subscription.
+	blockHash := common.HexToHash("0x1")
+	logStore.subscription.logs <- types.Log{
+		BlockHash: blockHash,
+	}
+
+	// Wait for the logs to be dispatched.
+	timeout, tCancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer tCancel()
+	err = e2eutils.WaitFor(timeout, 500*time.Millisecond, func() (bool, error) {
+		result := logStore.GetLogByBlockHash(blockHash)
+		return result[0].BlockHash == blockHash, nil
+	})
+	require.NoError(t, err)
+
+	// Quit the subscription.
+	logStore.Quit()
+}
+
+// TestLogStore_DispatchLogs_SubscriptionError tests the [DispatchLogs] method on the [logStore]
+// when the subscription returns an error.
+func TestLogStore_DispatchLogs_SubscriptionError(t *testing.T) {
+	query := ethereum.FilterQuery{}
+	client := newMockLogStoreClient()
+	log := testlog.Logger(t, log.LvlError)
+	logStore := NewLogStore(query, client, log)
+
+	// Subscribe to the logStore.
+	err := logStore.Subscribe()
+	require.NoError(t, err)
+
+	// Dispatch logs on the logStore.
+	go logStore.dispatchLogs()
+
+	// Send an error through the subscription.
+	client.sub.errorChan <- errors.New("test error")
+	time.Sleep(1 * time.Second)
```

Comment:
Need to avoid arbitrary sleeps - there's no guarantee that async tasks will complete in 1s - CI servers can be surprisingly busy and unpredictable. We'd either need the stub subscription to expose a channel that reports reconnections so we can use that to synchronise or we'd need to use a waiter to wait until the desired effect happens.